### PR TITLE
Gas Abstraction

### PR DESCRIPTION
**Pull request by petejkim**
Requires #300.

* Adds `increaseAllowance` and `decreaseAllowance` to mitigate ERC20 multi-withdrawal attack vulnerability `approve` has.
* Adds `transferWithAuthorization`, `approveWithAuthorization`, `increaseAllowanceWithAuthorization`, `decreaseAllowanceWithAuthorization` to enable ETH-less transactions.
* Adds `permit` – an alternative to approveWithAuthorization, provided for compatibility with the draft EIP2612 proposed by Uniswap. Differences [are laid out here](https://github.com/centrehq/centre-tokens/commit/3c69a144a028754766ad9f0962e73c9c2093499d#diff-a5901d56946b375959670d992772bae5R33).

### CODE REVIEW

**Comment by eztierney** [comment]
Slowly going through this, but in general looks good so far.

One thing that would be helpful is adding a doc in the `docs` directory explaining the new v2 features - including things like how to sign approvals, why we we have the `*WithAuthorization` methods and `permit`, etc.

I can puzzle most of it out from the code/comments but would be nice to have it all in 1 place for future integrators who are trying to build on these features.

File: test/v2/safeAllowance.behavior.ts [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        // check that allowance is now 0
+        expect((await fiatToken.allowance(alice, bob)).toNumber()).to.equal(0);
+        // check that Approval event is emitted
+        expect(log.event).to.equal("Approval");
+        expect(log.args[0]).to.equal(alice);
+        expect(log.args[1]).to.equal(bob);
+        expect(log.args[2].toNumber()).to.equal(0);
+      });
+
+      it("reverts if the decrease is greater than current allowance", async () => {
```
Comment (eztierney) [inline test/v2/safeAllowance.behavior.ts:120]: Also test for overflow?

**Comment by petejkim** [comment]
Improved test coverage. New code is all at 100% now.

```
-------------------------------|----------|----------|----------|----------|----------------|
File                           |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
-------------------------------|----------|----------|----------|----------|----------------|
 upgradeability/               |      100 |    78.57 |      100 |      100 |                |
  AdminUpgradeabilityProxy.sol |      100 |       80 |      100 |      100 |                |
  Proxy.sol                    |      100 |      100 |      100 |      100 |                |
  UpgradeabilityProxy.sol      |      100 |       75 |      100 |      100 |                |
 util/                         |      100 |      100 |      100 |      100 |                |
  ECRecover.sol                |      100 |      100 |      100 |      100 |                |
  EIP712.sol                   |      100 |      100 |      100 |      100 |                |
 v1.1/                         |      100 |      100 |      100 |      100 |                |
  FiatTokenV1_1.sol            |      100 |      100 |      100 |      100 |                |
  Rescuable.sol                |      100 |      100 |      100 |      100 |                |
 v1/                           |      100 |     96.3 |      100 |      100 |                |
  AbstractFiatTokenV1.sol      |      100 |      100 |      100 |      100 |                |
  Blacklistable.sol            |      100 |      100 |      100 |      100 |                |
  FiatTokenProxy.sol           |      100 |      100 |      100 |      100 |                |
  FiatTokenV1.sol              |      100 |    94.74 |      100 |      100 |                |
  Ownable.sol                  |      100 |      100 |      100 |      100 |                |
  Pausable.sol                 |      100 |      100 |      100 |      100 |                |
 v2/                           |      100 |      100 |      100 |      100 |                |
  AbstractFiatTokenV2.sol      |      100 |      100 |      100 |      100 |                |
  EIP712Domain.sol             |      100 |      100 |      100 |      100 |                |
  FiatTokenV2.sol              |      100 |      100 |      100 |      100 |                |
  GasAbstraction.sol           |      100 |      100 |      100 |      100 |                |
  Permit.sol                   |      100 |      100 |      100 |      100 |                |
-------------------------------|----------|----------|----------|----------|----------------|
All files                      |      100 |    94.44 |      100 |      100 |                |
-------------------------------|----------|----------|----------|----------|----------------|
```
